rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+
+## Required Approvers
+
+- Engineering: @jimbishopp
+- Product: @xinding33
+
+## What
+
+Users can change their permissions in a way that leaves them incapable of making further changes to the cluster - in effect, they lock themselves out.
+
+Actions that can cause this state:
+
+- user
+  - removal
+  - role unassigment
+- role
+  - edit
+- SSO
+  - connector removal
+  - user/group removal
+
+Those actions will cause a lock when:
+
+| action                 | initial state                                     | change                                                                                 |
+| ---------------------- | ------------------------------------------------- | -------------------------------------------------------------------------------------- |
+| user removal           | _admin_, the user with user deletion capabilities | user with user deletion capabilities deletes _admin_                                   |
+| user role removal      | _admin_                                           | _admin_ removes himself from a role that grants him user/role editing capabilities     |
+| role edit              | _admin_                                           | _admin_ edits own role to remove user/role editing capabilities                        |
+| SSO connector removal  | _SSO admin_                                       | _SSO admin_ removes SSO connector. _SSO admin_ logs out.                               |
+| SSO user/group removal | _SSO admin_                                       | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_ |
+
+Teleport already protects against a similar case in the following scenario:
+
+| action       | initial state | change                           |
+| ------------ | ------------- | -------------------------------- |
+| role removal | _admin_       | _admin_ tries to remove own role |
+
+The error message is displayed: _failed to delete role that still in use by a user. Check system server logs for more details_.
+
+There is one more case when a user can lock himself out:
+
+| initial state | change                                                                                    |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |
+
+However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.
+
+## Terminology
+
+- _admin_: local user with user/role editing capabilities
+- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)
+
+## Why
+
+When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
```

Comment:
```suggestion
When a user locks themselves out, they are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
```